### ISSUE TITLE
Microsoft has announced support for glTF in Paint 3D

### ISSUE THREAD

**Issue by superusercode** [KhronosGroup/glTF#1037]
As per https://blogs.windows.com/windowsexperience/2017/07/12/announcing-more-updates-to-paint-3d-magic-select-enhancements-drawing-tools/#VIIRzsIkuqOuJGB3.97 the [website](https://www.khronos.org/gltf) should be updated to recognize this.

**Comment by pjcozzi** [comment]
This is such fantastic news!

@donmccurdy do you think you can work this into your README.md updates?

@outofcontrol I also agree that it is worth an update to https://www.khronos.org/gltf

Also, please help spread the word on twitter: https://twitter.com/iamSBTron/status/885189708803854336

**Comment by superusercode** [comment]
Just as a note, Paint 3D has an online counterpart (pretty much like Sketchfab) that is called Remix3D. This platform will also have support for gltf thanks to Babylon.js.

**Comment by emackey** [comment]
I have several Windows 10 desktops, and I'm not on any "insider's build" or beta programs.  All of my stock Win10 machines have auto-updated themselves such that Paint 3D can import, edit, and export `*.glb` files.

There doesn't appear to be support for `*.gltf` files, just `*.glb`.  I'm guessing this is by design.  Those of us who maintain glTF tooling should probably shore up support for the binary flavor, it seems like this one more than the others is the one that's poised to go main-stream.

**Comment by CentaurMare** [comment]
Have you tried Paint 3D with a glb?

It doesn't properly import untextured models, it doesn't seem to import textured models at all.
Exports to glb fail in the validator.

I wrote a glb exporter for Sketchup, wanting to port my models across for painting in Paint 3D - the glb files validate and show correctly in Don's viewer. But opening in Paint 3D only one mesh is imported, and Paint 3D refuses to open any with textures.

**Comment by emackey** [comment]
Hmm, looks like I only tested one file, Avocado.glb, and that one works fine.  Only problem there is the "default" camera view, which Paint3D enforces fairly strongly, is the back of the Avocado.  I think this is opposite the default view of most of our other glTF viewers.

![avocadopaint1](https://user-images.githubusercontent.com/1235080/28252237-18684098-6a5d-11e7-9e29-e934963deda4.png)

Different sample models came from different converters and exporters.  I think Avocado came from the same system that did the Lantern and the Corset, which also appear to work.

The sample models that came through the COLLADA converter don't appear to load at all in Paint3D for me.

**Comment by CentaurMare** [comment]
Buggy loads, but not if you double-click the glb, only through the open menu.
But then, not all nodes/meshes are displayed, and some are positioned incorrectly.
![buggy](https://user-images.githubusercontent.com/30219662/28252391-53105d84-6ad5-11e7-9d17-720ed9c3449f.png)

**Comment by superusercode** [comment]
From what I understand, Windows 10's Insider program users have the latest version of Paint 3D. Various fixes may have been made in that version that are not on stable yet.

**Comment by cx20** [comment]
I tried to display glb file of [glTF-Sample-Model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0) repository with Paint 3D.
When a texture is used, the glb model generated by COLLADA2GLTF seems to be unable to display in Paint 3D yet.

Windows 10 version : 1703 (OS build 15063.483)
Paint 3D version : 2.1707.1017.0

## Simple models for testing individual features

|Model                                                                                                                   |generator                   |use texture|Paint 3D glb loading result              |
|:-----------------------------------------------------------------------------------------------------------------------|----------------------------|:---------:|-----------------------------------------|
|[Triangle Without Indices](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TriangleWithoutIndices)   |hand-crafted                |No         |model N/A                                |
|[Triangle](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Triangle)                                 |hand-crafted                |No         |model N/A                                |
|[Animated Triangle](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/AnimatedTriangle)                |hand-crafted                |No         |model N/A                                |
|[Animated Morph Cube](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/AnimatedMorphCube)             |glTF Tools for Unity        |No         |:white_check_mark: animation not support |
|[Animated Morph Sphere](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/AnimatedMorphSphere)         |glTF Tools for Unity        |No         |:white_check_mark: animation not support |
|[Simple Meshes](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/SimpleMeshes)                        |hand-crafted                |No         |model N/A                                |
|[Cameras](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Cameras)                                   |hand-crafted                |No         |model N/A                                |

## More complex models

|Model                                                                                                                   |generator                   |use texture|Paint 3D glb loading result              |
|:-----------------------------------------------------------------------------------------------------------------------|----------------------------|:---------:|-----------------------------------------|
|[Box](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Box)                                           |COLLADA2GLTF                |No         |:white_check_mark:                       |
|[Box Textured](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BoxTextured)                          |COLLADA2GLTF                |Yes        |:x:                                      |
|[Duck](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Duck)                                         |COLLADA2GLTF                |Yes        |:x:                                      |
|[Smiling Face](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/SmilingFace) vertex color problem     |COLLADA2GLTF                |Yes        |:x:                                      |
|[2 Cylinder Engine](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/2CylinderEngine)                 |COLLADA2GLTF                |No         |:x: No error, but not displayed          |
|[Reciprocating Saw](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/ReciprocatingSaw)                |COLLADA2GLTF                |No         |:x: No error, but not displayed          |
|[Gearbox Assy](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/GearboxAssy)                          |COLLADA2GLTF                |No         |:x: No error, but only partial           |
|[Buggy](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Buggy)                                       |COLLADA2GLTF                |No         |:x: No error, but only partial           |
|[Box Animated](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BoxAnimated)                          |COLLADA2GLTF                |No         |:white_check_mark: animation not support |
|[Cesium Milk Truck](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/CesiumMilkTruck)                 |COLLADA2GLTF                |Yes        |:x:                                      |
|[Rigged Simple](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/RiggedSimple)                        |COLLADA2GLTF                |No         |:white_check_mark: animation not support |
|[Rigged Figure](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/RiggedFigure)                        |COLLADA2GLTF                |No         |:white_check_mark: animation not support |
|[Walking Lady](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/WalkingLady)                          |COLLADA2GLTF                |No         |:white_check_mark: animation not support |
|[Cesium Man](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/CesiumMan)                              |COLLADA2GLTF                |Yes        |:x:                                      |
|[Monster](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Monster)                                   |COLLADA2GLTF                |Yes        |:x:                                      |
|[BrainStem](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BrainStem)                               |COLLADA2GLTF                |No         |:x: No error, but only partial           |
|[Virtual City](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/VC)                                   |COLLADA2GLTF                |Yes        |:x:                                      |

## PBR models

|Model                                                                                                                   |generator                   |use texture|Paint 3D glb loading result              |
|:-----------------------------------------------------------------------------------------------------------------------|----------------------------|:---------:|-----------------------------------------|
|[Avocado](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Avocado)                                   |glTF Tools for Unity        |Yes        |:white_check_mark:                       |
|[Barramundi Fish](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BarramundiFish)                    |glTF Tools for Unity        |Yes        |:white_check_mark:                       |
|[Boom Box](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/BoomBox)                                  |glTF Tools for Unity        |Yes        |:white_check_mark:                       |
|[Corset](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Corset)                                     |glTF Tools for Unity        |Yes        |:white_check_mark:                       |
|[Lantern](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Lantern)                                   |glTF Tools for Unity        |Yes        |:white_check_mark:                       |
|[MetalRoughSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/MetalRoughSpheres)               |COLLADA2GLTF with hand-edits|Yes        |model N/A                                |
|[NormalTangentTest](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/NormalTangentTest)               |COLLADA2GLTF with hand-edits|Yes        |model N/A                                |

## Further PBR models

|Model                                                                                                                   |generator                   |use texture|Paint 3D glb loading result              |
|:-----------------------------------------------------------------------------------------------------------------------|----------------------------|:---------:|-----------------------------------------|
|[Two Sided Plane](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/TwoSidedPlane)                     |VKTS glTF 2.0 exporter      |Yes        |model N/A                                |
|[Cube](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Cube)                                         |VKTS glTF 2.0 exporter      |Yes        |model N/A                                |
|[Animated Cube](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/AnimatedCube)                        |VKTS glTF 2.0 exporter      |Yes        |model N/A                                |
|[Suzanne](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Suzanne)                                   |VKTS glTF 2.0 exporter      |Yes        |model N/A                                |
|[Sci Fi Helmet](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/SciFiHelmet)                         |VKTS glTF 2.0 exporter      |Yes        |model N/A                                |

/cc @sbtron @bghgary

**Comment by CentaurMare** [comment]
@cx20, nice work!
It seems that the common denominator for the textured models is that the ones that work all have 
have normal maps and metallic maps. Maybe the normal maps are mandatory for Paint 3D?

A new model exported from Paint 3D fails validation...
```

        "uri": "Untitled.glb",
        "mimeType": "model/gltf-binary",
        "errors": {
            "BUFFER_VIEW_TARGET_OVERRIDE": [
                {
                    "type": "BUFFER_VIEW_TARGET_OVERRIDE",
                    "path": "#/images/0/bufferView",
                    "message": "Override of previously set bufferView target or usage. Initial: `VertexBuffer`, new: `Image`."
                },
                {
                    "type": "BUFFER_VIEW_TARGET_OVERRIDE",
                    "path": "#/images/1/bufferView",
                    "message": "Override of previously set bufferView target or usage. Initial: `VertexBuffer`, new: `Image`."
                },
                {
                    "type": "BUFFER_VIEW_TARGET_OVERRIDE",
                    "path": "#/images/2/bufferView",
                    "message": "Override of previously set bufferView target or usage. Initial: `VertexBuffer`, new: `Image`."
                },
                {
                    "type": "BUFFER_VIEW_TARGET_OVERRIDE",
                    "path": "#/images/3/bufferView",
                    "message": "Override of previously set bufferView target or usage. Initial: `VertexBuffer`, new: `Image`."
                }
            ]
        },

```

**Comment by pjcozzi** [comment]
@cx20 do you plan to track Paint3D as part of `gltf-test`?  That would be great.

**Comment by CentaurMare** [comment]
OK, I've figured out what is wrong with Paint 3D.

First of all, the name field is used to uniquely identity each material. A model without name fields defining materials will end up using the first material for every mesh.
Just naming the material anything is enough to separate the material, for example:

```
  "materials": [
    {
		"name" : "Fred",
      "pbrMetallicRoughness": {
        "baseColorFactor": [
          0.980392,
          0.921569,
          0.843137,
          1.0
        ],
        "metallicFactor": 0.0
      },
      "doubleSided": true
    },
    {
	"name" : "Wilma",
      "pbrMetallicRoughness": {
        "baseColorFactor": [
          1.0,
          0.0,
          0.0,
          1.0
        ],
        "metallicFactor": 0.0
      },
      "doubleSided": true
    }
  ],

```

Secondly, node names must be used and must be unique:

```
"nodes": [
    {
		"mesh": 0,
		"name": "Bart"
    },
    {
      "mesh": 1,
	  "name": "Homer"
    },
    {
      "mesh": 2,
	  "name": "Lisa"
	  
    },
    {
      "mesh": 3,
	  "name": "Marge"
    },
    {
	    "children": [
        0,
        1,
        2,
        3
      ],
	  "name": "Maggie"
    }
  ],

```

And finally, mesh names must be used, and must be unique:

```
  "meshes": [
    {
	  "name": "Bubbles",
      "primitives": [
        {
          "attributes": {
            "POSITION": 0,
            "NORMAL": 1
          },
          "indices": 2,
          "mode": 4,
          "material": 0
        }
      ]
    },
    {
	  "name": "Blossom",
      "primitives": [
        {
          "attributes": {
            "POSITION": 0,
            "NORMAL": 1
          },
          "indices": 3,
          "mode": 4,
          "material": 1
        }
      ]
    },
    {
	  "name": "Buttercup",
      "primitives": [
        {
          "attributes": {
            "POSITION": 0,
            "NORMAL": 1
          },
          "indices": 4,
          "mode": 4,
          "material": 0
        }
      ]
    },
    {
	  "name": "Professor Neutonium",
      "primitives": [
        {
          "attributes": {
            "POSITION": 0,
            "NORMAL": 1
          },
          "indices": 5,
          "mode": 4,
          "material": 1
        }
      ]
    }
  ],
```

**Comment by sbtron** [comment]
Thanks everyone for the detailed bug reports. We continue to push out updates that will improve support over time.

For models with textures failing to load – Its likely due to a bug where Paint incorrectly requires a metalrough texture if a baseColor texture is specified. 
@CentaurMare you shouldn't need to specify normal or other textures. Could you try just adding a metalroughtexture and see if that works? That will help verify its the same issue you are seeing.

I don't think there is any requirement for names to be unique etc. There might be something else that is causing that issue. Would it be possible to share the model where you were seeing that issue so we can investigate further?

Also sketchup exporter for glTF sounds very interesting is it posted somewhere?

**Comment by CentaurMare** [comment]
@sbtron,
I tried a model with with just metallicRoughnessTexture (along with baseColorTexture), and that works!

What I did with my test model (test.glb) was systematically modify the json payload (trial and error), until I got something that worked in paint 3d (test_x.glb). To do that I rewrote the glb incorporating the new json payload.

[test.zip](https://github.com/KhronosGroup/glTF/files/1154012/test.zip)

C# code for modifying the header:
```
static void Main(string[] args)
        {
            byte[] glb = System.IO.File.ReadAllBytes(@"test.glb");
            byte[] json = System.IO.File.ReadAllBytes(@"test.json");
            UInt32 header = BitConverter.ToUInt32(glb, 0);
            UInt32 version = BitConverter.ToUInt32(glb, 4);
            UInt32 length = BitConverter.ToUInt32(glb, 8);
            Console.WriteLine(header.ToString("x8"));
            Console.WriteLine(version.ToString("x8"));
            Console.WriteLine(length.ToString("x8"));

            UInt32 jsonLength = BitConverter.ToUInt32(glb, 12);
            UInt32 jsonHeader = BitConverter.ToUInt32(glb, 16);
            
            Console.WriteLine(jsonHeader.ToString("x8"));
            Console.WriteLine(jsonLength.ToString("x8"));

            UInt32 bufLength = BitConverter.ToUInt32(glb, 20 + (int)jsonLength);
            UInt32 bufHeader = BitConverter.ToUInt32(glb, 24 + (int)jsonLength);
            

            Console.WriteLine(bufHeader.ToString("x8"));
            Console.WriteLine(bufLength.ToString("x8"));

            List<byte> jsonArray = new List<byte>();
            jsonArray.AddRange(json);
            while (jsonArray.Count % 4 != 0)
                jsonArray.Add((byte)' ');
            UInt32 jsonLength2 = (UInt32)jsonArray.Count;

            List<byte> array = new List<byte>();
            array.AddRange(BitConverter.GetBytes(header));
            array.AddRange(BitConverter.GetBytes(version));
            length = jsonLength2 + 28 + bufLength;
            array.AddRange(BitConverter.GetBytes(length));
            array.AddRange(BitConverter.GetBytes(jsonLength2));
            array.AddRange(BitConverter.GetBytes(jsonHeader));
            array.AddRange(jsonArray);

            array.AddRange(BitConverter.GetBytes(bufLength)); 
            array.AddRange(BitConverter.GetBytes(bufHeader));

            for(int i = 0;i<bufLength;i++)
            {
                byte b = glb[28 + (int)jsonLength + i];
                array.Add(b);
            }
            System.IO.File.WriteAllBytes(@"test_x.glb", array.ToArray());
        }

```

I haven't posted my exporter anywhere at this time, as I haven't implemented matrices (I am currently multiplying all the geometry in the model by the combined matrix in sketchup during the export), and also haven't implemented shared meshes (components in Sketchup). So it is not quite ready for release.

Regarding Paint 3D, I'd be interested in your comments on my post here:
[https://forums.sketchup.com/t/update-on-paint-3d/48476](url)

Basically I was keen to have a workflow of export from Sketchup, paint in Paint 3D, import back into Sketchup.

**Comment by sbtron** [comment]
@CentaurMare  the pixelation issues you are seeing in the imported model are likely not related to the file inputs but due to down sampling in Paint. It should not have been this noticeable though. Would it be possible to share that sample file so we could investigate further?

**Comment by CentaurMare** [comment]
See attached:

[Berenice.zip](https://github.com/KhronosGroup/glTF/files/1158141/Berenice.zip)

It would be nice to know what Paint3D is actually doing, I initially expected it to only work with texture mapped images, but it seems to roll its own uv maps?

**Comment by CentaurMare** [comment]
@sbtron,

I have found another issue with the Paint3D implementation of glTF.

See attached.

[spheres.zip](https://github.com/KhronosGroup/glTF/files/1158695/spheres.zip)

I've been rewriting my Sketchup to binary gltf exporter, and looking for where I could make improvements. For example, previously I had one buffer view for every accessor.

One improvement I made was to have one mesh contain multiple materials (via the primitives array).
But, it appears Paint 3D ignores anything after the first primitive.

**Comment by pjcozzi** [comment]
@CentaurMare when your Sketchup exporter is ready, could you please open a pull request to a link to it to the list of glTF Tools in this repo: https://github.com/KhronosGroup/glTF/tree/donmccurdy-docs-tools-cleanup#gltf-tools (will be merged to master soon)

CC @donmccurdy

**Comment by CentaurMare** [comment]
@sbtron
It seems also that Paint 3D expects the texture specified for metallicRoughnessTexture to be the same dimensions as the baseColorTexture. I was hoping to embedded a 1x1 default texture in my Sketchup glTF exporter but this isn't going to work, so the only thing I can do at the moment is use the same texture specified for baseColorTexture as the metallicRoughnessTexture.

@pjcozzi, @donmccurdy 
For my Sketchup glTF exporter, I'm waiting on acceptance as a developer on the http://extensions.sketchup.com site.
Once I've been approved (assuming I get approved), I then need to submit the ruby code and documentation, which will then get validated by (one or more of) their team.

So the process is apparently quite slow, it may be a week or more.

Using that website means the code is encrypted but still usable in Sketchup, the alternative is to post it as open source.

**Comment by pjcozzi** [comment]
@CentaurMare the plugin news is exciting.  Please let us know when it is approved, we'd love to check it out and help promote your work!  Also, it would be great for the glTF ecosystem if it were open source and that would potentially also help you get some extra contributors to fix bugs, etc., but that is totally your call based on the goals of the goal.
